developer_docs/testing.md

```diff
@@ -0,0 +1,472 @@
+
+----
+For example, if you wanted to run just the SketchList test but also update the snapshot, you could do:
+```
+npm run test -- Sketchlist.test.js -u
+```
+
+Find more commands in the [Jest documentation](https://jestjs.io/docs/cli).
+
+## Why write tests
+- Good place to start if you're learning the codebase because it's harder to mess up production code
+- Benefits all future contributors by allowing them to check their changes for errors
+- Catches easy-to-miss errors
+- Lets you check your own work and feel more comfortable sumbitting PRs
+- Good practice for large projects
+- Many of the existing components don't have tests yet, and you could write one :-)
+
+## When to run tests
+
+When you make a git commit, the tests will be run automatically for you. 
+
+When you modify an existing component, it's a good idea to run the test suite to make sure it didn't make any changes that break the rest of the application. If they did break some tests, you would either have to fix a bug component or update the tests to match the new expected functionality.
+
+## Writing a test
+Want to get started writing a test for a new file or an existing file, but not sure how?
+### For React components
+(the below assumes we're using proposed folder structure 1)
+1. Make a new file in the ``__tests__`` folder that's directly adjacent to your file. For example, if ``example.jsx`` is in ``src/components``, then you would make a file called ``example.test.jsx`` in ``src/components/__tests__``
+2. Check if the component is connected to redux or not.
+3. If it is, see the redux section below on how to write tests for that.
+4. If it's not, see the section below on writing tests for unconnected components.
+
+### What to test
+For any type of component, you might want to consider testing:
+- User input results in the class's method being called. 
+  ```
+  //component is the return value of calling render()
+  const spy1 = jest.spyOn(component.instance(), 'func1');
+  act(() => {
+    fireEvent.click(screen.getByTestId("testid"));
+  });
+  expect(spy1).toHaveBeenCalledTimes(1);
+  ``` 
+- The text or divs that you expect to be on the page are actually there.
+- a previously saved snapshot of the HTML matches a snapshot taken during testing.
+- what else?? help!
+
+### For Redux action creators or reducers
+See the [redux section](#Testing-Redux) below :)
+
+### Troubleshooting
+1. Check if the component makes any API calls. If it's using axios, jest should already be set up to replace the axios library with a mocked version; however, you may want to [mock](https://jestjs.io/docs/mock-function-api#mockfnmockimplementationoncefn) the axios.get() function with your own version so that GET calls "return" whatever data makes sense for that test. 
+
+    ```
+    axios.get.mockImplementationOnce(
+      (x) => Promise.resolve({ data: 'foo' })
+    );
+    ```
+You can see it used in the context of a test [here](../client/modules/IDE/components/SketchList.test.jsx).
```

Comment:
For screen reader accessibility reasons, it's best practice to not have links with the text "here" or  "click here", but to instead describe what the link goes to. See [What's wrong with using "Click Here" links?](https://webaccess.berkeley.edu/ask-pecan/click-here)

Reply:
fixed, good to know! Thanks for letting me know~